src/audio/sources/flac.cpp

```diff
@@ -0,0 +1,221 @@
+// This file is part of playd.
+// playd is licensed under the MIT licence: see LICENSE.txt.
+
+/**
+ * @file
+ * Implementation of the FlacAudioSource class.
+ * @see audio/audio_source.hpp
+ */
+
+#include <cassert>
+#include <cstdint>
+#include <cstdlib>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <sstream>
+#include <string>
+
+#include <FLAC++/decoder.h>
+
+#include "../../errors.hpp"
+#include "../../messages.h"
+#include "../sample_formats.hpp"
+#include "../audio_source.hpp"
+#include "flac.hpp"
+
+/* static */ std::unique_ptr<AudioSource> FlacAudioSource::Build(
+                const std::string &path)
+{
+	return std::unique_ptr<AudioSource>(new FlacAudioSource(path));
+}
+
+FlacAudioSource::FlacAudioSource(const std::string &path)
+    : AudioSource(path), buffer()
+{
+	auto err = this->init(path);
+	if (err != FLAC__STREAM_DECODER_INIT_STATUS_OK) {
+		throw FileError("flac: can't open " + path + ": " +
+		                FlacAudioSource::InitStrError(err));
+	}
+
+	// We need to decode some to get the sample rate etc.
+	// Because libflac is callback-based, we have to spin on waiting for
+	// the sample rate to materialise.
+	do {
+		this->process_single();
+	} while (this->get_sample_rate() == 0);
+
+	Debug() << "flac: sample rate:" << this->SampleRate() << std::endl;
+	Debug() << "flac: bytes per sample:" << this->BytesPerSample()
+	        << std::endl;
+	Debug() << "flac: mono bytes per sample:"
+	        << (this->get_bits_per_sample() / 8) << std::endl;
+	Debug() << "flac: channels:" << (int)this->ChannelCount() << std::endl;
+	Debug() << "flac: playd format:" << (int)this->OutputSampleFormat()
+	        << std::endl;
+}
+
+FlacAudioSource::~FlacAudioSource()
+{
+	this->finish();
+}
+
+/* static */ std::string FlacAudioSource::InitStrError(int err)
+{
+	switch (err) {
+		case FLAC__STREAM_DECODER_INIT_STATUS_OK:
+			return "success";
+		case FLAC__STREAM_DECODER_INIT_STATUS_UNSUPPORTED_CONTAINER:
+			return "libflac missing support for container";
+		case FLAC__STREAM_DECODER_INIT_STATUS_INVALID_CALLBACKS:
+			return "internal error: bad callbacks";
+		case FLAC__STREAM_DECODER_INIT_STATUS_MEMORY_ALLOCATION_ERROR:
+			return "out of memory";
+		case FLAC__STREAM_DECODER_INIT_STATUS_ERROR_OPENING_FILE:
+			return "fopen() failed";
+		case FLAC__STREAM_DECODER_INIT_STATUS_ALREADY_INITIALIZED:
+			return "internal error: already initialised?";
+		default:
+			return "unknown error";
+	}
```

Comment:
Yeah, not seeing a map here to be honest, unless it was an old-school array indexed by constant (which is somewhat flimsy because we don't control the constants)